### PR TITLE
New Concordance adapter

### PR DESCRIPTION
**Pull request by irina060981**
For https://github.com/alpheios-project/client-adapters/issues/4

It creates TextQuoteSelector, WordUsageExample, Author and TextWork
Also it has additional methods - convertToJson  and readObjects for Homonym, Lexeme, Lemma, Feature, Translation (it is from wordlist/worditem branch)

I checkout from feat-304-word-list (to use TextQuoteSelector as a parent for WordUsageExample for further using in annotations model) - that is why it is here.

### CODE REVIEW

File: src/w3c/word-usage-example.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  /**
+  * Creates WordUsageExample object from jsonObj, homonym, author, textWork and link from the adapter config
+  * @param {Object} jsonObj - json object from concordance api
+  * @param {Homonym} homonym - source homonym object
+  * @param {Author} author - source author object, could be undefined
+  * @param {TextWork} textWork - source textWork object, could be undefined
+  * @param {String} sourceLink - sourceTextUrl from the adapter config file
+  * @returns {WordUsageExample}
+  */
+  static readObject (jsonObj, homonym, author, textWork, sourceLink) {
```
Comment (balmas) [inline src/w3c/word-usage-example.js]: I think this also it too tightly coupled to the PHI concordance service's output format. Please parse the JSON of the service output there, and explicitly construct a WordUsageExample .  See more in the client-adapters PR.

File: src/w3c/word-usage-example.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import TextQuoteSelector from './text-quote-selector'
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/w3c/word-usage-example.js]: Yes, I agree completely here. I have made them parent-child only with the aim not to loose it relations in future. The final structure here should consider all the relations between WordItem, TextQuoteSelector, WordUsageExample and Homonym. For now there is not enough information for it (from my point ov view of course)

File: src/w3c/author.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  * @param {String} urn - string identificator in special format, for example 'urn:cts:latinLit:phi0959'
+  * @param {Object} titles - has the following format { languageCode: title }
+  * @param {Object} abbreviations - has the following format { languageCode: abbreviation }
+  * @returns {Author}
+  */
+  constructor (urn, titles, abbreviations) {
+    this.urn = urn
+    this.titles = titles
+    this.abbreviations = abbreviations
+    this.ID = this.extractIDFromURN()
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/w3c/author.js]: Done

File: src/w3c/text-work.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  * @param {Object} titles - has the following format { languageCode: title }
+  * @param {Object} abbreviations - has the following format { languageCode: abbreviation }
+  * @returns {TextWork}
+  */
+  constructor (author, urn, titles, abbreviations) {
+    this.urn = urn
+    this.titles = titles
+    this.author = author
+    this.abbreviations = abbreviations
+    this.ID = this.extractIDFromURN()
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/w3c/text-work.js]: Done

File: src/w3c/text-work.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  */
+  static get defaultIDPrefix () {
+    return 'phi'
+  }
+
+  /**
+  * Method returns title in default language or (if not exists) it returns first available title
+  * @returns {String}
+  */
+  get title () {
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/w3c/text-work.js]: Done

File: src/w3c/text-work.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      return Object.values(this.titles)[0]
+    }
+    return null
+  }
+
+  /**
+    * Method returns abbreviation in default language or (if not exists) it returns first available abbreviation
+    * @returns {String}
+    */
+  get abbreviation () {
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/w3c/text-work.js]: Done

File: src/w3c/text-work.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    return null
+  }
+
+  /**
+  * Method returns TextWork for given jsonObj (from concordance API)
+  * @param {Author} author - author of the textWork
+  * @param {Object} jsonObj - json object with data of the TextWork
+  * @returns {TextWork}
+  */
+  static create (author, jsonObj) {
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/w3c/text-work.js]: Done

File: src/w3c/word-usage-example.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  * @returns {String}
+  */
+  get fullCit () {
+    let res = ''
+    if (this.author) {
+      res = this.author.title
+      if (this.textWork) {
+        res = res + ' ' + this.textWork.title
+      } else {
+        if (this.cit && this.cit.split('.') && this.cit.split('.').length >= 2) {
```
[the remark this answers is not in the document]
Comment (irina060981) [reply src/w3c/word-usage-example.js]: @balmas , Agree, I also think that in future when there will be another cit formats we could create get/set cit property and one fullCit calculated property, that is used for showing o the screen.

File: src/w3c/word-usage-example.js (hunk 1 above)
Comment (irina060981) [reply src/w3c/word-usage-example.js]: Done

File: src/w3c/author.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+  get abbreviation () {
+    if (this.abbreviations[Author.defaultLang]) {
+      return this.abbreviations[Author.defaultLang]
+    } else if (Object.values(this.abbreviations).length > 0) {
+      return Object.values(this.abbreviations)[0]
+    }
+    return null
+  }
+
+  /**
```
[the remark this answers is not in the document]
Comment (balmas) [reply src/w3c/author.js]: I think it makes sense to have methods to convert to/from json in each model object, but they should be the mirror of each other. I.e. the data-model class should have a method to serialize to JSON and to deserialize from that same JSON format. But if model objects are created from other sources, then the code for constructing the object from JSON should live close to the source. Does that make sense?

[the remark this answers is not in the document]
Comment (irina060981) [reply src/w3c/author.js]: You mean that we should have to/from JSON methods for saving/uploding to IndexedDB in data-models. And parsing data from source inside client-adapters.
I think it makes sense :)

**Pull request merged**